### PR TITLE
Check return code

### PR DESCRIPTION
**Pull request by lucafuji**
check return code of the spawn process

### CODE REVIEW

**Comment by bolkedebruin** [comment]
@lucafuji can you please squash your commits?

**Comment by lucafuji** [comment]
@bolkedebruin 
Just squash all the commits

**Comment by mistercrunch** [comment]
Ok I took the time to read this. `BaseJob.run` is not meant to be overridden, `_execute` is, and the way you did it leads to copy pasta duplication, which is bad because the same logic needs to be maintained in different places.

It think that line 887 to 889 are (almost) the only lines of code that need to be added to  `_execute` after the while loop.

**Comment by lucafuji** [comment]
@mistercrunch 
Yep, I agree there are some copy and paste code
However anyway you still need to overwrite the run() function
because when the _execute() exits, even it marks itself as failed,
the run function will still override the job status to be success

**Comment by mistercrunch** [comment]
Then we should change `BaseJob.run` to `try:` and record failure when `_execute` raises, and make `LocalJob._execute` raise when it has a return code.

**Comment by lucafuji** [comment]
I changed according to comments.
Basically I add a on_failure callback when some exception happens during execution and implemented only for LocalTaskJob otherwise it will mark it self as success in the database

File: airflow/jobs.py [hunk 1]
```diff
@@ -863,10 +871,23 @@ def _execute(self):
         while return_code is None:
             self.heartbeat()
             return_code = self.process.poll()
+        if return_code != 0:
+            msg = ("LocalTaskJob process exited with non zero status "
+                   "{return_code}".format(**locals()))
+            logging.error(msg)
```
Comment (mistercrunch) [inline airflow/jobs.py]: Shouldn't there be a `raise AirflowExcetion(msg)` here instead?

File: airflow/jobs.py [hunk 2]
```diff
@@ -153,11 +158,14 @@ def run(self):
         self.id = id_
 
         # Run
-        self._execute()
+        try:
+            self._execute()
+            self.state = State.SUCCESS
+        except Exception as e:
+            self.on_failure(e)
```
Comment (mistercrunch) [inline airflow/jobs.py:167]: `self.state = State.FAILED` should be here as well as `logging.exception(e)`

Comment (lucafuji) [inline airflow/jobs.py:167]: Just changed according to comments

File: airflow/jobs.py (hunk 1 above)
Comment (lucafuji) [inline airflow/jobs.py]: Just changed according to comments

**Comment by lucafuji** [comment]
Weird, from the checks I see none are related to my diffs

File: airflow/utils.py [hunk 3]
```diff
@@ -115,6 +115,12 @@ def runnable(cls):
             None, cls.FAILED, cls.UP_FOR_RETRY, cls.UPSTREAM_FAILED,
             cls.SKIPPED, cls.QUEUED]
 
+    @classmethod
+    def failed(cls):
+        return [
```
Comment (mistercrunch) [inline airflow/utils.py:120]: nit: would fit on one line

File: airflow/utils.py [hunk 4]
```diff
@@ -115,6 +115,12 @@ def runnable(cls):
             None, cls.FAILED, cls.UP_FOR_RETRY, cls.UPSTREAM_FAILED,
             cls.SKIPPED, cls.QUEUED]
 
+    @classmethod
+    def failed(cls):
```
Comment (mistercrunch) [inline airflow/utils.py:119]: `unsuccessful` ?

**Comment by artwr** [comment]
@lucafuji, aside from the nits, this still looks useful. If you want to still contribute, you could rebase and we could check it against current tests. A unit test would be welcome. If we do not hear in 5 days, I will probably close as abandoned.

**Comment by lucafuji** [comment]
@artwr Sure, I will rebase from master and submit another diff soon

**Pull request closed without merging**